content/en/platform/corda/5.0/developing-applications/getting-started/utxo-advanced-ledger-extensions/advanced-design.md

```diff
@@ -0,0 +1,413 @@
+* On chainable state(s) updating, the previous state pointer of every created chainable state must not be null.
+* On chainable state(s) updating, the previous state pointer of every created chainable state must be pointing to exactly one consumed chainable state, exclusively.
+
+```kotlin
+public final class Update extends ChainableContractUpdateCommand<ExampleChainableState> {
+  
+   @NotNull
+   
+    public Class<ExampleChainableState> getContractStateType() {
+     return ExampleChainableState.class;
+  }
+  
+   @Override
+  
+    protected void onVerify(@NotNull final UtxoLedgerTransaction transaction) {
+    
+    // Verify additional Update constraints
+  }
+}
+```
+
+The `ChainableContractDeleteCommand` supports deleting existing chainable states and will verify the following constraint:
+
+* On chainable state(s) deleting, at least one chainable state must be consumed.
+
+```kotlin
+
+    public final class Delete extends ChainableContractDeleteCommand<ExampleChainableState> {
+  
+    @NotNull
+    
+     public Class<ExampleChainableState> getContractStateType() {
+      return ExampleChainableState.class;
+  }
+  
+   @Override
+  
+    protected void onVerify(@NotNull final UtxoLedgerTransaction transaction) {
+    
+    // Verify additional Delete constraints
+    
+  }
+}
+```
+
+## Designing a Chainable Contract
+
+A chainable contract can be implemented by extending the `ChainableContract` class; for example:
+
+```kotlin
+public final class ExampleChainableContract extends ChainableContract {
+  
+   @Override
+  
+   public List<Class<? extends ChainableContractCommand<?>>> getPermittedCommandTypes() {
+      return List.of(Create.class, Update.class, Delete.class);
+      
+  }
+}
+```
+
+## Fungible API
+
+Module: fungible
+
+Package: com.r3.corda.ledger.utxo.fungible
+
+The fungible API provides the component model for designing fungible states and contracts. Fungible states represent states that have a scalar numeric quantity, and can be split, merged and mutually exchanged with other fungible states of the same class. Fungible states represent the building blocks for states like tokens.
+
+## Designing a Fungible State
+
+A fungible state can be implemented by implementing the `FungibleState<T>` interface; for example:
+
+```kotlin
+public final class ExampleFungibleState extends FungibleState<NumericDecimal> {
+  
+  @NotNull
+  
+  private final NumericDecimal quantity;
+  
+  public ExampleFungibleState(@NotNull final NumericDecimal quantity) {
+    this.quantity = quantity;
+  }
+  
+  @NotNull
+  
+  public NumericDecimal getQuantity() {
+    return quantity;
+  }
+  
+  @NotNull
+  
+  public List<PublicKey> getParticipants() {
+    return List.of(...);
+  
+  }
+  
+  
+  @Override
+  
+   public boolean isFungibleWith(@NotNull final FungibleState<NumericDecimal> other) {
+   return this == other || other instanceof ExampleFungibleState // && other fungibility rules.
+   
+  }
+}
+```
```

Comment:
```suggestion
```kotlin
public final class ExampleFungibleState extends FungibleState<NumericDecimal> {  
  @NotNull
  private final NumericDecimal quantity;
  
  public ExampleFungibleState(@NotNull final NumericDecimal quantity) {
    this.quantity = quantity;
  }
  
  @NotNull
  public NumericDecimal getQuantity() {
    return quantity;
  }
  
  @NotNull
  public List<PublicKey> getParticipants() {
    return List.of(...);
  }
  
  @Override
  public boolean isFungibleWith(@NotNull final FungibleState<NumericDecimal> other) {
    return this == other || other instanceof ExampleFungibleState // && other fungibility rules.
  }
}
```
```